### ISSUE TITLE
Privileges not propagated through "r <= s" annotations

### ISSUE THREAD

**Issue by manopapad** [StanfordLegion/legion#421]
This snippet:
```
import 'regent'
task bar(r : region(ispace(int2d), float))
where reads writes(r) do
  for idx in r do
    r[idx] += 1
  end
end
task foo(r : region(ispace(int2d), float),
         p1_c1 : region(ispace(int2d), float))
where reads writes(r), p1_c1 <= r do
  bar(p1_c1)
end
task main()
  var r = region(ispace(int2d,{6,6}), float)
  var cs1 = ispace(int2d,{2,1})
  var p1 = partition(equal, r, cs1)
  for c1 in cs1 do
    foo(r, p1[c1])
  end
end
regentlib.start(main)
```
compiles, but produces a runtime failure:
```
[0 - 7fcc19e3f700] {5}{runtime}: [error 298] LEGION ERROR: Parent task foo (ID 4) of task bar (ID 7) does not have a region requirement for region (4,1,1) as a parent of child task's region requirement index 0 (from file /home/manolis/sw/legion/runtime/legion/legion_tasks.cc:2343)[error 298] 
For more information see:
http://legion.stanford.edu/messages/error_code.html#error_code_298
```
If I inline `bar` inside `foo`:
```
import 'regent'
task foo(r : region(ispace(int2d), float),
         p1_c1 : region(ispace(int2d), float))
where reads writes(r), p1_c1 <= r do
  for idx in p1_c1 do
    p1_c1[idx] = 1
  end
end
task main()
  var r = region(ispace(int2d,{6,6}), float)
  var cs1 = ispace(int2d,{2,1})
  var p1 = partition(equal, r, cs1)
  for c1 in cs1 do
    foo(r, p1[c1])
  end
end
regentlib.start(main)
```
I instead get a compile-time assertion fail:
```
/home/manolis/sw/legion/language/src/regent/codegen.t:341: assertion failed!
stack traceback:
	[C]: in function 'assert'
	/home/manolis/sw/legion/language/src/regent/codegen.t:341: in function 'fn'
	src/asdl.lua:18: in function 'fn'
	src/asdl.lua:18: in function 'map'
	/home/manolis/sw/legion/language/src/regent/codegen.t:1098: in function '__ref'
	/home/manolis/sw/legion/language/src/regent/codegen.t:1237: in function 'write'
	/home/manolis/sw/legion/language/src/regent/codegen.t:8616: in function 'fn'
	src/asdl.lua:18: in function 'block'
	/home/manolis/sw/legion/language/src/regent/codegen.t:7829: in function 'fn'
	src/asdl.lua:18: in function 'block'
	/home/manolis/sw/legion/language/src/regent/codegen.t:9515: in function 'compile_thunk'
	/home/manolis/sw/legion/language/src/regent/std_base.t:1288: in function 'complete'
	/home/manolis/sw/legion/language/src/regent/std.t:3918: in function </home/manolis/sw/legion/language/src/regent/std.t:3916>
	/home/manolis/sw/legion/language/src/regent/std.t:3920: in function 'compile_tasks_in_parallel'
	/home/manolis/sw/legion/language/src/regent/std.t:4035: in function 'start'
	bug-privileges-subset-1.rg:17: in main chunk
```

**Comment by lightsighter** [comment]
Can you make a Legion Spy log for one of these runs? Build regent with the '--spy' flag.

**Comment by manopapad** [comment]
Here's the spy log produced when I run the first snippet: [spy_0.log](https://github.com/StanfordLegion/legion/files/2319811/spy_0.log).

**Comment by elliottslaughter** [comment]
This is probably a compiler bug. However, I don't think it can be what you think it is. `bar` takes one and only one region requirement. `foo` takes two, but only one has privileges. The other region is `NO_ACCESS`, if it is being passed at all.

The real bug is that the compiler ought to see that `p1_c1 <= r`, and so `p1_c1` should be the root of privileges. I don't know why this could be broken, since we've been doing it forever, but it must be where the bug is because I can't see any other way to get that error.

**Comment by lightsighter** [comment]
Ok, yeah I was looking at 'foo', but Legion Spy says that foo had the following region requirements:

0. READ-WRITE on 'r'
1. NO_ACCESS on 'p1_c1'

**Comment by elliottslaughter** [comment]
The immediate problem is fixed and the first test above should now pass. However, in the process of fixing this, I realized that there is an issue similar to #289 which can occur here. Specifically, it's possible for different fields to derive privileges from different parents, and for this to be incompatible with the callee's expectations. For now, I just take the first parent with non-zero privileges that I find, and if the user splits privileges like this then they'll get a runtime error.

**Comment by elliottslaughter** [comment]
@manopapad Should be fixed as of cd5a4fc. Please confirm.

**Comment by manopapad** [comment]
Fixed!
